Signal/src/util/OWSBackupExportJob.m

```diff
@@ -3,64 +3,253 @@
 //
 
 #import "OWSBackupExportJob.h"
+#import "OWSBackupIO.h"
 #import "OWSDatabaseMigration.h"
+#import "OWSSignalServiceProtos.pb.h"
 #import "Signal-Swift.h"
 #import <SignalServiceKit/NSData+Base64.h>
 #import <SignalServiceKit/NSDate+OWS.h>
 #import <SignalServiceKit/OWSBackgroundTask.h>
-#import <SignalServiceKit/OWSBackupStorage.h>
 #import <SignalServiceKit/OWSError.h>
 #import <SignalServiceKit/OWSFileSystem.h>
 #import <SignalServiceKit/TSAttachment.h>
 #import <SignalServiceKit/TSAttachmentStream.h>
 #import <SignalServiceKit/TSMessage.h>
 #import <SignalServiceKit/TSThread.h>
 #import <SignalServiceKit/Threading.h>
-#import <SignalServiceKit/YapDatabaseConnection+OWS.h>
-#import <YapDatabase/YapDatabase.h>
 
 NS_ASSUME_NONNULL_BEGIN
 
-NSString *const kOWSBackup_ExportDatabaseKeySpec = @"kOWSBackup_ExportDatabaseKeySpec";
+@class OWSAttachmentExport;
 
+@interface OWSBackupExportItem : NSObject
+
+@property (nonatomic) OWSBackupEncryptedItem *encryptedItem;
+
+@property (nonatomic) NSString *recordName;
+
+// This property is optional and is only used for attachments.
+@property (nonatomic, nullable) OWSAttachmentExport *attachmentExport;
+
+// This property is optional.
+//
+// See comments in `OWSBackupIO`.
+@property (nonatomic, nullable) NSNumber *uncompressedDataLength;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+#pragma mark -
+
+@implementation OWSBackupExportItem
+
+- (instancetype)initWithEncryptedItem:(OWSBackupEncryptedItem *)encryptedItem
+{
+    if (!(self = [super init])) {
+        return self;
+    }
+
+    OWSAssert(encryptedItem);
+
+    self.encryptedItem = encryptedItem;
+
+    return self;
+}
+
+@end
+
+#pragma mark -
+
+// Used to serialize database snapshot contents.
+// Writes db entities using protobufs into snapshot fragments.
+// Snapshot fragments are compressed (they compress _very well_,
+// around 20x smaller) then encrypted.  Ordering matters in
+// snapshot contents (entities should we restored in the same
+// order they are serialized), so we are always careful to preserve
+// ordering of entities within a snapshot AND ordering of snapshot
+// fragments within a bakckup.
+//
+// This stream is used to write entities one at a time and takes
+// care of sharding them into fragments, compressing and encrypting
+// those fragments.  Fragment size is fixed to reduce worst case
```

Comment:
I've called it stream because it has a stream-like interface and because eventually it might be nice to use `NSStream`, but I consider that a nice-to-have.  Using `NSStream` would only offer slight benefits - a single file instead of multiple fragments.